docs/howto/danceable-lyrics-input-tool/DanceableLyrics/backend/ayx_plugins/danceable_lyrics.py

```diff
@@ -26,13 +29,34 @@
 class DanceableLyrics(PluginV2):
     """Concrete implementation of an AyxPlugin."""
 
+    @staticmethod
+    def _validate_datasets_dir(datasets_dir: Path) -> None:
+        """Ensure that the datasets directory exists, and contains all the expected files."""
+        if not datasets_dir.is_dir():
+            raise WorkflowRuntimeError("Bad path")
+
+        filenames = [
+            "artists.csv",
+            "tracks.csv",
+            "r_track_artist.csv",
+            "genius_song_lyrics.csv",
+            "audio_features.csv"
+        ]
+
+        existing_files = [((datasets_dir / filename).is_file(), datasets_dir / filename) for filename in filenames]
```

Comment:
List comprehension syntax creates a new list. Although this isn't important for this tool. It uses extra memory. Also issues with heap allocation and garbage collection (pending on Python runtime internals).

Furthermore a list is being created with a tuple of `(bool, full_path)`. Only `bool` is being used.

The following example doesn't create a new list.
```
is_file_iterable = map(lambda f: (datasets_dir / filename).is_file(), filenames)
if not all(is_file_iterable):
   ...
```

it is the equivalent more or less of this:
```
def all_exists():
   for file in filenames:
      exists = (datasets_dir / file).is_file()
      if not exists:
         return False
   return True
```

Reply:
the bool gets used in the check, but the full path gets used in the error message. I wanted to return the set of files that don't exist - when I first ran the tool, I ran into an issue with the truncated dataset, because the tool didn't register `genius_song_lyrics_TRUNCATED.csv` as `genius_song_lyrics.csv`

Still, I think the map syntax should be cleaner, so I'll try using that - standby

Reply:
I totally missed that. Could simply filter/collect on not exists. 
```
iter = filter(files, not_exists)
list = ', '.join(iter)
if not list:
   raise ... "Expected files not found: " + list
```

Reply:
Just realized it's way easier to check the list length after the filter - no idea where my head was at when I first wrote this

Reply:
You can just simply join and check the string length. It'll be empty if all files exist.